Fix elections-phragmen and proxy issue

kianenigma opened the pull request:
Closes https://github.com/paritytech/substrate/issues/7030
polkadot companion: https://github.com/paritytech/polkadot/pull/1719

📕 Detailed description of the PR is [here](https://github.com/paritytech/substrate/pull/7040#issuecomment-730578716)

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -483,17 +488,11 @@ decl_module! {
-				// remove the reporter.
-				Self::do_remove_voter(&reporter, false);
+				// remove the defunct target while unreserving their bond.
+				Self::do_remove_voter(&target, true);
```
So maybe we can do some kind of refund here, but I think that would open a new attack vector: 

The proxy can set a wrong vote, report it, and it will be for free. 

All in all, it is bit unfortunate that all modules need now be proxy-aware. The pallet was safe for a runtime without proxy and now it is not. Maybe we can come up with different and better ways to deal with such issues.

gavofyork replied:
> The proxy can set a wrong vote

No - voting should require a deposit on-chain equal to the resources that the vote takes. If the proxy sets a wrong vote, they can tie up funds from the main account into a deposit, but that's fine as the main account can remove the proxy and then free up the vote.

We don't have a concept of "defunct voter" any more. All voting is paid for and is allowed to remain on-chain until the voter (or their proxy) chooses to remove it, even if it is "defunct" (i.e. doesn't actually contributes toward any tallies).

The weight associated with this function should be done normally (i.e. it's enough to ensure that if it's called incorrectly (and results in a no-op), then the computation is paid for).

kianenigma replied:
Yes I first recommended that maybe this _cleanup_ operation can be made a free transaction if it was successful because it is doing the chain some favour, but as you said given the vote this is not really a favour anymore, thus okay as-is.  

> Crucially, we need to change vote from a fixed deposit to one that properly recognises the costs associated with the vote.

Addressed. 

> This is wrong - they don't.

Well, I can't see how we could have had the same vulnerability without the concept of proxies. 

For example, I was thinking: Ideally we could allow the proxy module to configure each proxy with some call indices that can be allowed (instead of the current `enum ProxyType` which is more high level and only distinguishes certain operations). Then, the person who is delegating a proxy has fine grained control over which calls their proxy can dispatch and can (nay -- _should_) prohibit the proxy from calling into such functions that can drain their bond. 

In that case, if they happen to allow it, it is no longer our responsibility and they've messed up their proxy configurations.

gavofyork replied:
Well, I would say that it didn't really strictly adhere to game-theoretic principles that we eventually developed elsewhere in the codebase (i.e. notably that any additional state should require a deposit according to the additional storage/computation it implies), and this was done even before we had this particular attack vector.

If you ensure that the actor themself always pays for their actions, then you're in good territory. The issue here was that "the chain" kind of paid for the repurcussions of the `vote` action as evidenced by the chain being willing to slash&repatriate funds in order to incentivize its cleanup. This should not be necessary as long as the voter is paying their way properly with a deposit.

In general, we should be very careful about repatriating funds. In hindsight, this is probably quite a big footgun and we might even want to deprecate it entirely.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -460,7 +466,6 @@ decl_module! {
 			let target = T::Lookup::lookup(defunct.who)?;
 
 			ensure!(reporter != target, Error::<T>::ReportSelf);
-			ensure!(Self::is_voter(&reporter), Error::<T>::MustBeVoter);
```
Note that before because we had to do (potential) slashing the origin (reporter) must have been a voter. Now it can be anyone.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -194,8 +194,14 @@ pub trait Trait: frame_system::Trait {
+	/// creating a gigantic number of votes.
+	type VotingBondBase: Get<BalanceOf<Self>>;
+
+	/// The amount of bond that need to be locked for each vote (32 bytes)
```
```suggestion
	/// The amount of bond that need to be locked for each vote (32 bytes).
```

kianenigma commented:
Weight update report. Note that we only care about DB stuff and variables. These are native numbers on my machine.


### `vote`

- before: `#[ weight  = 50 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(4, 2)]`
- now: 
```
(131493000 as Weight)
			.saturating_add(DbWeight::get().reads(5 as Weight))
			.saturating_add(DbWeight::get().writes(2 as Weight))
```
- comments: `set_lock` is both read and write of locks, and it was mis-counted before. Also before, we always submitted `MAX_VOTES`, now we have a variable `v`, so just more accurate. 

### `remove_voter`

- before: `#[weight = 35 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(1, 2)]`
- now: 
```
(90000000 as Weight)
			.saturating_add(DbWeight::get().reads(2 as Weight))
			.saturating_add(DbWeight::get().writes(2 as Weight))
```
- comments: `set_lock` is both read and write of locks, and it was mis-counted before. 

### `report_defunct_voter_incorrect`

- before: 
```
Weight::from(defunct.candidate_count).saturating_mul(2 * WEIGHT_PER_MICROS)
			.saturating_add(Weight::from(defunct.vote_count).saturating_mul(19 * WEIGHT_PER_MICROS))
			.saturating_add(T::DbWeight::get().reads_writes(6, 3))
```
- now: 
```
(0 as Weight)
			.saturating_add((1664000 as Weight).saturating_mul(c as Weight))
			.saturating_add((42696000 as Weight).saturating_mul(v as Weight))
			.saturating_add(DbWeight::get().reads(4 as Weight))
```
- comments: The changes are expected. An invalid report is now a no-op. See the new refund scheme.

### `report_defunct_voter_incorrect`

- before: 
```
Weight::from(defunct.candidate_count).saturating_mul(2 * WEIGHT_PER_MICROS)
			.saturating_add(Weight::from(defunct.vote_count).saturating_mul(19 * WEIGHT_PER_MICROS))
			.saturating_add(T::DbWeight::get().reads_writes(6, 3))
```
- now: 
```
(0 as Weight)
			.saturating_add((1613000 as Weight).saturating_mul(c as Weight))
			.saturating_add((40652000 as Weight).saturating_mul(v as Weight))
			.saturating_add(DbWeight::get().reads(6 as Weight))
			.saturating_add(DbWeight::get().writes(3 as Weight))
```
- comments: Excellent. 


### `submit_candidacy`

- before: 
```
#[weight =
	(35 * WEIGHT_PER_MICROS)
	.saturating_add(Weight::from(*candidate_count).saturating_mul(375 * WEIGHT_PER_NANOS))
	.saturating_add(T::DbWeight::get().reads_writes(4, 1))
]
```
- now: 
```
(169289000 as Weight)
			.saturating_add((351000 as Weight).saturating_mul(c as Weight))
			.saturating_add(DbWeight::get().reads(3 as Weight))
			.saturating_add(DbWeight::get().writes(1 as Weight))
```
- comments: Previous read count was wrong because we count `candidates.decode_len()` and `candidates` twice. 

### `renounce_candidacy`

- before: 
```
#[weight =  match *renouncing {
	Renouncing::Candidate(count) => {
		(18 * WEIGHT_PER_MICROS)
		.saturating_add(Weight::from(count).saturating_mul(235 * WEIGHT_PER_NANOS))
		.saturating_add(T::DbWeight::get().reads_writes(1, 1))
	},
	Renouncing::Member => {
		46 * WEIGHT_PER_MICROS +
		T::DbWeight::get().reads_writes(2, 2)
	},
	Renouncing::RunnerUp => {
		46 * WEIGHT_PER_MICROS +
		T::DbWeight::get().reads_writes(1, 1)
	}
}]
```
- now: 
```
// candidate
(106589000 as Weight)
			.saturating_add((216000 as Weight).saturating_mul(c as Weight))
			.saturating_add(DbWeight::get().reads(1 as Weight))
			.saturating_add(DbWeight::get().writes(1 as Weight))
// member 
(190000000 as Weight)
			.saturating_add(DbWeight::get().reads(3 as Weight))
			.saturating_add(DbWeight::get().writes(4 as Weight))
// runners-up
(56000000 as Weight)
			.saturating_add(DbWeight::get().reads(1 as Weight))
			.saturating_add(DbWeight::get().writes(1 as Weight))

```
- comments: previous r/w for member renounce was wrong, because we forgot that we also call into `T::ChangeMambers` which alters some state in the council module.

### `remove_member_without_replacement`

- before: 
```
T::MaximumBlockWeight::get()
```
- comments: Nothing else to do. Stays the same for now.

### `remove_member_with_replacement`

- before: 
```
50 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(3, 2)
```
- now: 
```
(111000000 as Weight)
			.saturating_add(DbWeight::get().reads(4 as Weight))
			.saturating_add(DbWeight::get().writes(5 as Weight))
```
- comments: Similarly, calls into T::ChangeMember were ignored before. 

### `remove_member_wrong_refund`

- before: 
```
6 * WEIGHT_PER_MICROS + T::DbWeight::get().reads_writes(1, 0)
```
- now: 
```
(19000000 as Weight)
			.saturating_add(DbWeight::get().reads(1 as Weight))
```
- comments: All good.

kianenigma commented:
Okay -- weights also re-done and pushed.

gavofyork commented:
@kianenigma how about that polkadot PR?

kianenigma commented:
This is changing the amount of deposit needed for each vote. Should we write a migration code and update the deposit of all previous voters as well? I guess yes.

kianenigma commented:
Runtime note: All accounts participating in the council governance as voter will be subject to a change in their reserved deposit, without taking any action.

shawntabrizi commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -248,18 +254,23 @@ benchmarks! {
 		)?;
 
-		// all the bailers go away.
+		// all the bailers go away. NOTE: we can simplify this. There's no need to cerate all these
```
```suggestion
		// all the bailers go away. NOTE: we can simplify this. There's no need to create all these
```

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -234,11 +230,17 @@ decl_storage! {
 		/// can never enter this vector and is always implicitly assumed to be a candidate.
 		pub Candidates get(fn candidates): Vec<T::AccountId>;
+
+		/// Should be used in conjunction with `on_runtime_upgrade` to ensure an upgrade is executed
+		/// once, even if the code is not removed in time.
+		pub NeedsMigration get(fn needs_migrations): bool = true;
```
This does not seem very generic. Should probably follow the Version Enum pattern we have used before

shawntabrizi replied:
Also I think for sanity, it should be `HasMigrated`, where the default value is false, and is set to true when done.

kianenigma replied:
Yeah agreed.

kianenigma replied:
Moved to enum.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -325,12 +327,68 @@ decl_module! {
 
+		fn on_runtime_upgrade() -> Weight {
+			if Self::needs_migrations() {
+				let mut consumed_weight = 0;
```
I think it is fine to just take a full block for a runtime upgrade operation.

shawntabrizi replied:
Can you also move this into its own function which is then called by `on_runtime_upgrade`?

This may allow us to keep the code around a little longer and allow other users to trigger the upgrade on their own time.

kianenigma replied:
thought so too, idk, it looked easy enough to do it accurately as well. Either way it is fine since it is a one time thing.

shawntabrizi replied:
Looking at it a bit more carefully, you can immediately deprecate `LagacyVotingBond`.

Then create a function:

```
pub fn migrate_deposit<T>(legacy_voting_bond: BalanceOf<T>) {
    ...
}
```

Then call this function from the runtime (not the pallet): https://github.com/paritytech/substrate/pull/7040

And pass the old voting bond in that way. This would avoid yet another breaking change for the pallet users.

kianenigma replied:
> Then call this function from the runtime (not the pallet): #7040

Can you explain this a bit more?

shawntabrizi replied:
We can trigger some runtime upgrade logic from the runtime `lib.rs` rather than from the pallet. See here:

https://github.com/paritytech/polkadot/pull/1636

kianenigma replied:
Sweet.

shawntabrizi reviewed (changes requested):
In general this looks good. Although I think my suggestion for refactoring the upgrade code is needed.

gavofyork commented:
> This is changing the amount of deposit needed for each vote. Should we write a migration code and update the deposit of all previous voters as well? I guess yes.

the amount held on deposit (i.e. the old bond) should be recorded properly.

i don't think snatching more deposit from anyone is a good idea.

gavofyork commented:
and `LegacyVotingBond` can just be a config trait that implements `Get`

kianenigma commented:
> the amount held on deposit (i.e. the old bond) should be recorded properly.

> i don't think snatching more deposit from anyone is a good idea.

Hmm the only dangerous edge case that I see is where an account might get locked out of their account because they can exactly pay for the new deposit and no more, thus no more transaction are possible. Otherwise I think this is a small change and most accounts will stay unaffected with a very small change to their reserved balance. 

Alternative approach would be a holy pain; we'd need to record which votes have the old deposit and which have a new one (by potentially migrating all votes from `(stake, target)` to `(stake, target, old_or_new)`), and then upon `remove_voter` and other operations, check this flag and operate based on the voter having an old or new deposit. Then the`LegacyVotingBond ` need to also stay around potentially until we know all voters have re-submitted their votes and are migrated to the new bond. 

I am a bit unsure. This adds a sizeable amount of complication to the module, and will need another day or two of work to be done right. Preventing it from the get go will save us all some time in the future. 

We can do this alternatively: Make sure an account is not locked out by ensuring that we only update their deposit, if they have a threshold of unreserved and unlocked balance left. Then if they don't like this, they are free to `remove_voter` and get all of the bond back (which is possible because they surely don't have all their bond reserved and can pay for the transaction fee).

gavofyork commented:
I would do what we do with all modern deposit-taking pallets and record the deposit taken.

Migration should simply record each voting account with the deposit taken as the old deposit. This constant can be discarded along with the migration code.

Any new voting activity should record the new value.

kianenigma commented:
Applied both requested chnages: 

1. Deposit is now recorded. Migration only notes the previous deposit. 
2. LegacyVotingBond is not kept the runtime being migrated needs to provide it. 

Needs a thorough re-review.

kianenigma commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -28,10 +28,18 @@ use crate::Module as Elections;
 
 type Lookup<T> = <<T as frame_system::Trait>::Lookup as StaticLookup>::Source;
 
+macro_rules! whitelist {
```
Later on I should move these to a common file.

kianenigma commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -152,62 +167,59 @@ fn clean<T: Trait>() {
-		// User account seed
-		let u in 0 .. 1000 => ();
-	}
+	_ {}
```
Is there a way to get rid of this? Seemingly the macro fails without it.

shawntabrizi replied:
No, it is required for the moment. But in general, I am a fan of removing this aspect of the macro all together. I think it actually may cause more problems than it solves in bigger picture.

kianenigma commented on frame/elections-phragmen/src/default_weights.rs:
```diff
@@ -0,0 +1,87 @@
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+
+use frame_support::weights::{Weight, constants::RocksDbWeight as DbWeight};
```
need to re-generate these once in the companion as well.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -113,6 +114,42 @@ type BalanceOf<T> =
+		if <Module<T>>::pallet_storage_version() == StorageVersion::V1 {
+
+			<StorageKeyIterator<T::AccountId, (BalanceOf<T>, Vec<T::AccountId>), Twox64Concat>>::new(
+				b"PhragmenElection",
```
I'd really like to not rely on Strings here and get the module name from `decl_storage`. There's probably a good way for it.

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -350,53 +403,83 @@ decl_module! {
-			if !Self::is_voter(&who) {
-				T::Currency::reserve(&who, T::VotingBond::get())
+			// Reserve bond.
+			let Voter { votes: old_votes, deposit: old_deposit, .. } = <Voting<T>>::get(&who);
```
super critical code block to review for sanity. ⚠️ ☢️⚠️ ☢️⚠️ ☢️

kianenigma commented:
Migration tested on both polkadot and kusama as such:

https://github.com/paritytech/substrate-debug-kit/blob/146a1cb80610346040054fe9df7f172374b504d8/remote-ext-elections-phragmen/src/main.rs#L124-L149

gavofyork commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -350,53 +408,83 @@ decl_module! {
+				None
+			} else {
+				// Old voter.
+				match votes.len().cmp(&old_votes.len()) {
```
This is bad.

Instead of messing with `len`s, just calculate the new amount to be reserved, and find the difference with what has currently been reserved. If it's greater, reserve the difference. If it's less, unreserve the differrence.

This way even if the calculation/constants for `deposit_factor_of` change between `vote`s, you'll still ensure that the latest prices are the ones that the user pays.

If you don't do that, then suppose the prices go up 2x between votes: A user could make 10 votes before, replace them with 5 of them later, be fully unreserved, yet still have 5 votes.

kianenigma replied:
right this logic was made with the assumption of _not_ storing the deposit.

kianenigma replied:
Addressed, thanks!

joepetrowski commented on bin/node/runtime/src/lib.rs:
```diff
@@ -559,6 +559,8 @@ impl pallet_collective::Trait<CouncilCollective> for Runtime {
 
 parameter_types! {
 	pub const CandidacyBond: Balance = 10 * DOLLARS;
+	pub const VotingBondBase: Balance = 1 * DOLLARS;
+	pub const VotingBondFactor: Balance = 50 * CENTS;    // per 32 bytes on-chain
 	pub const VotingBond: Balance = 1 * DOLLARS;
```
Looks like `VotingBond` is unused?

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -153,6 +207,15 @@ pub trait WeightInfo {
 	fn remove_member_wrong_refund() -> Weight;
 }
 
+#[derive(Encode, Decode, Eq, PartialEq)]
+/// Storage version.
```
```suggestion
/// Storage version.
#[derive(Encode, Decode, Eq, PartialEq)]
```

I think habit is that doc goes first then attributes, so that attributes don't get lost before big docs

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -346,42 +421,67 @@ decl_module! {
+						debug_assert!(_remainder.is_zero());
+					},
+				};
+				Some(T::WeightInfo::vote_update(votes.len() as u32))
```
I think the benchmark only benchmarks for when new_deposit == old_deposit, but here it is used in any condition

kianenigma replied:
Great point, this needs a small fix.

kianenigma replied:
addressed, working now on updating the weight values.

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -346,42 +421,67 @@ decl_module! {
+			let new_deposit = Self::deposit_of(votes.len());
+			let Voter { votes: old_votes, deposit: old_deposit, .. } = <Voting<T>>::get(&who);
+			let maybe_refund = if old_votes.is_empty() {
+				// First time voter. Get full deposit.
```
I usually prefer not to admit that old_votes.is_empty is same as first time voter, because maybe in the future some mecanism can remove some voter or something and this property won't be respected anymore.
* why not using the fact that old_deposit is 0 ?
* or use Voting as `Option<Voter>`
* or just remove this special case which just saves almost nothing compare to the case `new_deposit > old_deposit`

kianenigma replied:
I like the third option the best 👍

kianenigma replied:
Addressed.

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -115,6 +115,47 @@ type BalanceOf<T> =
+				<Voting<T>>::module_prefix(),
+				b"Voting",
+			)
+			.for_each(|(who, (stake, votes))| {
```
sidenote: maybe we should change https://substrate.dev/rustdocs/v2.0.0/frame_support/storage/trait.IterableStorageMap.html#tymethod.translate to return the number of value translated, it would help also for weight

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -115,6 +115,47 @@ type BalanceOf<T> =
+				"🏛 pallet-elections-phragmen: {} voters migrated to V2PerVoterDeposit.",
+				count,
+			);
+			Weight::max_value()
```
would be good to have a migration test IMHO

kianenigma replied:
Looking into it, and I already did one here: https://github.com/paritytech/substrate/pull/7040#issuecomment-694190522

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -115,6 +115,47 @@ type BalanceOf<T> =
+	/// Migrate from the old legacy voting bond (fixed) to the new one (per-vote dynamic).
+	///
+	/// Will only be triggered if storage version is V1.
+	pub fn migrate_to_recorded_deposit<T: Trait>(old_deposit: BalanceOf<T>) -> Weight {
```
The benefit of this is that we can get rid of `type VotingBond` (see node/runtime).

You are right about the storage and it should either way have a `build(|_| StorageVersion::V2PerVoterDeposit)`. Correct?

thiolliere replied:
but we can get rid of VotingBond even if we do the migration with on_runtime_upgrade

thiolliere replied:
but we can get rid of VotingBond even if we do the migration with on_runtime_upgrade

kianenigma replied:
No in that case the value of old deposit is different in polkadot and kusama and we won't be able to know it from within this pallet (see the companion as well and how we are passing the old value of deposit into this function as argument. )

thiolliere replied:
Ah indeed you're right

apopiak replied:
Also want to note that it seems like a good practice to extract it out into a function for e.g. testability.

shawntabrizi commented:
/benchmark runtime pallet pallet_elections_phragmen

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

shawntabrizi commented:
/benchmark runtime pallet pallet_elections_phragmen

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

kianenigma commented:
I believe all comments are addressed, all weights are updated.

thiolliere reviewed (dismissed):
looks good to me

shawntabrizi commented:
@kianenigma how about the default weight? The bot right now only updates the weights for the node.

shawntabrizi commented:
Nit optimization:

https://github.com/paritytech/substrate/blob/c79b522a11bbc7b3cf2f4a9c0a6627797993cb79/frame/elections-phragmen/src/lib.rs#L613

This could be `append`

shawntabrizi reviewed (changes requested):
This is not finished AFAICT.

The candidate bond is not tracked anywhere in storage and you risk over/under reserving/slashing

thiolliere commented:
> This is not finished AFAICT.
> 
> The candidate bond is not tracked anywhere in storage and you risk over/under reserving/slashing

I thing we can do it in another PR
(sorry to have pointed wrongly to this PR, I mixed voter deposit tracking and candidate deposit tracking)

shawntabrizi commented:
@thiolliere i dont think we should do it in another PR though? this seems like a pretty real issue where we can easily mess up our tracking of reserved currencies.

apopiak commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -115,6 +115,47 @@ type BalanceOf<T> =
+	///
+	/// Will only be triggered if storage version is V1.
+	pub fn migrate_to_recorded_deposit<T: Trait>(old_deposit: BalanceOf<T>) -> Weight {
+		if <Module<T>>::pallet_storage_version() == StorageVersion::V1 {
```
We could try out the pallet storage version (see https://github.com/paritytech/substrate/pull/7208) here.
Note, though, that the version would be `None` for now.

apopiak replied:
mmh, just tried this and it seems complicated. Probably to better to wait with using `PalletVersion` until it is set in storage.

thiolliere replied:
> Probably to better to wait with using PalletVersion until it is set in storage?

this PR won't be part of v26 anymore thus we can make use of them now.

kianenigma commented:
This PR is blocked on a long track of fixes. Will come back to it hopefully soon.

kianenigma commented:
PR has been updated with a number of changes. I will summarize everything here: 

### Recorded Deposit for Everyone

Both voters and candidates have their deposits recorded. Consequently, all members/runners-up also get a recorded deposit. 

I've added tests to make sure we behave sanely when the bonds change on the fly. For example, if you are a candidate and by the time you become a member the bond value changes, your recorded bond is the amount that you had when you were a candidate.

### No more defunct voter.

The functionality of Defunct voter is removed all-together. A root call is added to clean the useless voters just in case, but I don't expect it be needed anytime soon, because we record appropriate deposits based on the count of the votes.

### Transaction version change. 

This removes one call and adds a new one. Transaction version need to be bumped. ⚠️

### Preventing old errors.

Changed the name of the notorious `compute_members_diff` to `compute_members_diff_sorted ` to prevent further mistakes. 

### Revamping all docs.

Updated all docs to the latest state of affairs. Also, I got rid of most detailed `<weight>` comments and only left the high level notes there. With the modern benchmarking system, I doubt if we really care about these manual enumerations of DB read/writes anymore. 

### Accurate `on_initialize` weight 

I've made some effort into properly weighing the on-chain phragmen as well here, thus making the weight of the `on_initialize` accurate in case of a new term.

### Deposit Factor Amount 

I am a bit uncertain about the amount of deposit factor that should be kept. This should account for 32bytes of data on-chain, and the weight that is being incurred to the election. WIP.

kianenigma commented:
> Nit optimization:
> 
> https://github.com/paritytech/substrate/blob/c79b522a11bbc7b3cf2f4a9c0a6627797993cb79/frame/elections-phragmen/src/lib.rs#L613
> 
> This could be `append` right?

No were inserting into a certain index, not to the very end.

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -116,6 +124,131 @@ type BalanceOf<T> =
+				count += 1;
+			});
+
+			PalletStorageVersion::put(StorageVersion::V2RecordedDeposit);
```
is there a reason for splitting all the migrations ? also the doc should state that this writes V2 to storage as other migration can't be triggered after this one as passed.

kianenigma replied:
This part is not done yet, I want to use the crate version stuff.

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -633,17 +730,15 @@ decl_module! {
-			Self::remove_and_replace_member(&who).map(|had_replacement| {
-				let (imbalance, _) = T::Currency::slash_reserved(&who, T::CandidacyBond::get());
-				T::KickedMember::on_unbalanced(imbalance);
+			Self::remove_and_replace_member(&who, true).map(|had_replacement| {
```
sidenote: I feel this is kind of weird syntax, why not:
```rust
			let had_replacement = Self::remove_and_replace_member(&who, true)?;

			Self::deposit_event(RawEvent::MemberKicked(who.clone()));

			if !had_replacement {
				// if we end up here, we will charge a full block weight.
				Self::do_phragmen();
			}

			// no refund needed.
			Ok(None.into())
```

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -656,93 +751,135 @@ decl_module! {
-				members_with_stake.binary_search_by(|(ref m, ref _s)| m.cmp(&replacement))
+	/// ### Note
+	///
+	/// This function attempts to preserve the prime. If the removed members is not the prime, it is
+	/// set again via [`Trait::ChangeMembers` ] to prevent it being wiped.
```
If I understand correclty the prime is reset when Members are changed in order to preserve the prime in case the implementation of ChangeMembers trait only refer to the prime as an index inside the member set ?

I worry this specific usage of the trait can be easily missed as it isn't written anywhere else as far as I see.
This requirements could be written either:
* in the doc of `ChangeMembers`
* or in the associated type ChangeMembers saying that on top of ChangeMembers assumption the trait is used with this constraint.

kianenigma replied:
> in case the implementation of ChangeMembers trait only refer to the prime as an index inside the member set?

Hmm I didn't think about this at all.. yeah this should be documented. 

But still my usage should be safe, I assume? I am first setting the new members and then set the prime again. Even if an index has changed, then the new prime should be correct. 

I recall that we had such problems in kusama: if a members renounces, then we lose the prime until the next election. This kinda sucks.

thiolliere replied:
yes implementation is fine, I just worry those assumption gets expected somewhere else and then broke because ppl missed them.

thiolliere replied:
I'm finally ok, I feel this function is slightly overdoing, but that shouldn't lead ppl to think this property is ensured by election-phragmen usages nor ChangeMembers.

thiolliere replied:
ah I finally understand where i'm wrong. change_member_sorted reset the prime, thus it needs to be set again. OK all good then

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -656,93 +751,135 @@ decl_module! {
+				.map(|x| x.who.clone())
+				.collect::<Vec<_>>();
+			let outgoing = &[who.clone()];
+			let current_prime = T::ChangeMembers::get_prime().unwrap_or_default();
```
this is wrong we don't want this function to set prime to default when there is no prime in the first place.

kianenigma replied:
Great point

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -805,34 +955,16 @@ impl<T: Trait> Module<T> {
-		}
-	}
+		// remove storage, lock and unreserve.
+		<Voting<T>>::remove(who);
```
note: you could use `take` direclty

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -289,210 +466,158 @@ decl_error! {
-			defunct.candidate_count,
-			defunct.vote_count,
-		)]
-		fn report_defunct_voter(
```
yeah, I noted the transaction version change [here](https://github.com/paritytech/substrate/pull/7040#issuecomment-730578716). That should be enough.

apopiak commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -127,95 +127,88 @@ type NegativeImbalanceOf<T> =
+		let current_version = <Module<T> as GetPalletVersion>::current_version();
+		let maybe_storage_version = <Module<T> as GetPalletVersion>::storage_version();
+
+		if current_version == PalletVersion::new(3, 0, 0) {
```
This will disallow running this migration for future versions meaning that people that don't upgrade before the next change to the phragmen pallet will not be able to run this.
I think you should only match on the storage version here.

kianenigma replied:
Hmm good point, but I wonder if that can create any sort of footgun in an edge case where we accidentally execute a migration twice?

kianenigma replied:
Why I am worried about for example is if someone is trying to do multiple upgrades at once. In my restricted scheme, they cannot shoot themselves in the foot. In yours, I think they can?

apopiak replied:
For that we could set the version that you're intending to migrate **to** after the migration was successful.

kianenigma replied:
Can we set the version? The version will always be set to what is noted in Cargo.toml after a runtime upgrade afaik.

kianenigma replied:
Let's elaborate: 

Imagine we have another migration from 3.0.0 -> 4.0.0 (that also depends on the current storage layout introduced in 3.0.0).  So the migration in current PR is migration `M3` and the one introduced hypothetically in the future is `M4`. Also, assume, once we add M4 and bump to 4.0.0 we keep M3 in the code base. 

Now, someone who last had 2.0.0 want to upgrade to 4.0.0. If both migrations are written as you suggest (only assert on storage version) they can use the crate in version 4.0.0  but apply M3. This will apply this migration for them, but will write 4.0.0. in their storage version, which means that they can never apply M4. 

I think it all boils down to keeping migrations in the code base. If we remove the migration introduced in this PR upon introducing the next migration, we are safe and I can do what you suggest. Else, I prefer keeping it strict with asserting on both versions.

apopiak replied:
Ok, but this leads me to think we should not write the current version to storage when `on_runtime_upgrade` is present and instead leave that to the migration so we can avoid your scenario.
If the migration defines the version that it migrates to your scenario works out just fine. Every migration loads the version from storage, checks that it is the expected one, migrates, and then sets the version that it migrated to.
@bkchr If you agree with this scheme I think we might have to adjust the `on_runtime_upgrade`.

apopiak replied:
@kianenigma Your idea is that people pulling in this migration will either use it as is (in the context of a v3 pallet) or will modify it to fit their context (e.g. in a v4 pallet)?

kianenigma replied:
> Ok, but this leads me to think we should not write the current version to storage when on_runtime_upgrade is present and instead leave that to the migration so we can avoid your scenario.


Maybe `on_runtime_upgrade` of each module can return `Option<PalletVersion>`, and you get where I am going with this.

But this makes things complicated. I am fine with keeping the current scheme. The only hassle now is for external teams: if you are 3 major versions (i.e. 3 migrations behind) you cannot do them all at once. You need to do them one by one. But this hassle will protect them actually, so it is good.

thiolliere replied:
why not write version 3_0_0 in storage when this migration is successful, and remove this if condition ?

I would rather keep writing after on_runtime_upgrade automatically, this prevents more errors that it causes.

kianenigma replied:
I see that the main dilemma here is: 

1. Should this migration be executed as a `fn on_runtime_upgrade()` inside the pallet? 
2. Should this migration be executed as an additional upgrade directly given to executive? 

I assumed that in all cases the automatic write happens, while it is not the case. Automatic write happens only in case 1. 

Based on the assumption that the upgrade will be called from executive I can remove the additional `if` and fo a manual write.

kianenigma replied:
Honestly though, I think this migration should live inside the `fn on_runtime_upgrade()` of the pallet, and have all the needed if-conditions to make sure it is executed only when it is a jump from 2_0_0 to 3_0_0. Given this, the code can always live in substrate and allows chains that are multiple versions behind to also update them one by one.

thiolliere replied:
yes ideally it is better for it to live inside the on_runtime_upgrade, but how can we give the information about old deposits ?
Maybe as associated type on Config ? then this is bloating Config :-/ so I don't know what is best.

OnRuntimeUpgrade executes first CustomOnRuntimeUpgrade define at runtime level then pallets OnRuntimeUpgrade which writes into the storage the pallet version after their execution.

kianenigma replied:
> Maybe as associated type on Config ? then this is bloating Config :-/ so I don't know what is best.

Even worse about this is that now we bump to 3_0_0 because the change is breaking, and if we add new configs just for migration, we only can clean them if we bump to 4_0_0 because the removal itself will be a breaking change (I am doing my best here to adhere to semver, which is kinda hard to do in pallets.)

apopiak replied:
> I see that the main dilemma here is:
> 
>     1. Should this migration be executed as a `fn on_runtime_upgrade()` inside the pallet?
> 
>     2. Should this migration be executed as an additional upgrade directly given to executive?
> 
> 
> I assumed that in all cases the automatic write happens, while it is not the case. Automatic write happens only in case 1.

I don't think that's right. I thought [this code](https://github.com/paritytech/substrate/blob/f884296f7436916909025f8b43c4bbf3e60e4c60/frame/support/src/dispatch.rs#L1349-L1358) means that the version is set for all pallets.

kianenigma commented:
Detailed migration testing done here: https://github.com/paritytech/substrate-debug-kit/blob/migrate-substrate-7040/laboratory/remote-ext-elections-phragmen/src/main.rs#L129

kianenigma commented on frame/elections-phragmen/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "pallet-elections-phragmen"
-version = "2.0.0"
+version = "3.0.0"
```
@gnunicorn Can you elaborate a bit on how this works?

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -195,29 +326,40 @@ pub trait Trait: frame_system::Trait {
 		config(members): Vec<(T::AccountId, BalanceOf<T>)>;
 		build(|config: &GenesisConfig<T>| {
 			let members = config.members.iter().map(|(ref member, ref stake)| {
-				// make sure they have enough stake
```
Given that you now correctly track user's deposit, I don't see why this is a requirement. Just have the initial members have zero stake.

I think it may even be weird for people that genesis will try to reserve some value.

kianenigma replied:
kk -- genesis members will vote for themselves, with _**no deposit**_ and _**no locks**_. Kind of a fishy situation but I see also your point.

thiolliere replied:
Why initial members wouldn't need to have a correct deposit and locks ? I prefer to panic on genesis and having a more sane state than the opposite.

shawntabrizi replied:
It is the same thing as what we do with a lot of genesis state, like indices for example. All genesis indices have zero deposit. Its strange to me that genesis will be doing things to manipulate locks and balances, but maybe it is fine.

kianenigma replied:
I don't have a strong opinion on this so I will keep it as it is. 

My personal preference is what @thiolliere recommends, being accurate. 
But I think overall it is better for substrate pallets to be consistent and behave similarly. Given the indices example, not reserving makes more sense.

thiolliere replied:
ok then I'm fine with any ways.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -289,210 +434,158 @@ decl_error! {
+		/// we consider the common case of placing more votes. In other two case, we refund.
 		/// # </weight>
-		#[weight = T::WeightInfo::vote(votes.len() as u32)]
+		#[weight = T::WeightInfo::vote_more(votes.len() as u32)]
```
It's my opinion that this should be `vote_more.max(vote_less).max(vote_equal)`

It is basically zero cost, and provides more customization and doesnt make assumptions about the weight of these functions.

wheresaddie replied:
yeap I agree w Shawn, important to answer questions not create them.

kianenigma replied:
Not sure if I follow: `vote_more()` and `vote_less()`have additional state access (and arguably the same but that's a different story). If someone submits the same amount of votes, why should they pay for more expensive variant? doing a `.max()` is a backtrack to a more primitive version of weighing in which we just assume we always execute the most expensive path if the function. 

Even if we are to change, then it can stay as `#[weight = T::WeightInfo::vote_more(votes.len() as u32)]` without any of the refunds. `vote_more` is basically the max of all of them.

shawntabrizi replied:
> vote_more is basically the max of all of them.

I get that, but this is ultimately an assumption and something that may be configurable from the outside. Your initial weight here should always be the max of all possible paths, and you should use a simple max comparison to make that determination imo

kianenigma replied:
I will do as you recommend for now but I am not fully convinced yet.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -289,210 +434,158 @@ decl_error! {
+					// Must unreserve a bit.
+					let to_unreserve = old_deposit - new_deposit;
+					let _remainder = T::Currency::unreserve(&who, to_unreserve);
+					debug_assert!(_remainder.is_zero());
```
is this needed? I don't see a benefit...

kianenigma replied:
I have a tendency to sprinkle these invariants into the code as debug_assertions and they have saved my arse in the past, so preferably keeping them.

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -633,116 +698,160 @@ decl_module! {
+				// In both cases, we will change more weight than need. Refund and abort.
 				return Err(Error::<T>::InvalidReplacement.with_weight(
 					// refund. The weight value comes from a benchmark which is special to this.
 					//  5.751 µs
```
```suggestion
```

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -633,116 +698,160 @@ decl_module! {
 		}
 
-		/// What to do at the end of each block. Checks if an election needs to happen or not.
+		/// Clean all voters who are defunct (i.e. the do not serve any purpose at all). The deposit
```
```suggestion
		/// Clean all voters who are defunct (i.e. they do not serve any purpose at all). The deposit
```

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -633,116 +698,160 @@ decl_module! {
+		/// # <weight>
+		/// The total number of voters and those that are defunct must be provided as witness data.
+		/// # </weight>
+		#[weight = 0]
```
?

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -633,116 +698,160 @@ decl_module! {
+			let removed = members.remove(remove_index);
+			if slash {
+				let (imbalance, _remaining) = T::Currency::slash_reserved(who, removed.deposit);
+				debug_assert_eq!(_remaining, 0u32.into());
```
same question about if this is needed, and what purpose it is for.

Also would be better to do `_remaining.is_zero()`

shawntabrizi commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -633,116 +698,160 @@ decl_module! {
+			}))
+		})
+		.map(|maybe_replacement| {
+			let member_ids = Self::members()
```
maybe a nit, but here you query the overlay again, when def you could probably keep the members list in memory

shawntabrizi reviewed (approved):
I have some feedback, but generally looks right.

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -116,6 +126,118 @@ type BalanceOf<T> =
+
+		if current_version == PalletVersion::new(3, 0, 0) {
+			match maybe_storage_version {
+				Some(storage_version) if storage_version == PalletVersion::new(2, 0, 0) => {
```
chain on substrate 2.0 don't have version 2.0.0 in storage (storage version wasn't introduced yet), maybe we should do `<=` instead then. Or document it.

kianenigma commented on bin/node/runtime/src/lib.rs:
```diff
@@ -554,7 +554,10 @@ impl pallet_collective::Trait<CouncilCollective> for Runtime {
 parameter_types! {
 	pub const CandidacyBond: Balance = 10 * DOLLARS;
-	pub const VotingBond: Balance = 1 * DOLLARS;
+	// 1 storage item created, key size is 32 bytes, value size is 16+16.
+	pub const VotingBondBase: Balance = deposit(1, 64);
```
I also worry a bit about being outdated, but this is the pattern that is used in polkadot as well and I learned from that. I think for now it is good enough. 

As for the accurate key size, I am trying to lean what was intended in the implementation of `const fn deposit`. Looking at the usage, it seems like the function is intended to used without taking the first two_64 key into account. 

For example, given this storage: 

https://github.com/paritytech/substrate/blob/f884296f7436916909025f8b43c4bbf3e60e4c60/frame/multisig/src/lib.rs#L127-L130

We have deposit: 

https://github.com/paritytech/polkadot/blob/8c1a5a5b3e353154d8efa4daf78a25f199a3c16f/runtime/kusama/src/lib.rs#L708-L709

Which seems to take into account the secondary key of the double map (32 byte for blake2_128_concat). 

Similarly, for proxies that is just a normal map, we have

https://github.com/paritytech/polkadot/blob/8c1a5a5b3e353154d8efa4daf78a25f199a3c16f/runtime/kusama/src/lib.rs#L783-L784

So I think here the correct value here is `deposit(1, 32 + 1);`. The length is always below 16 and will fit into 1 byte of SCALE compact.

thiolliere replied:
I'm surprised proxy deposit base is `8` and not `32+8`.

But then I'm ok.

kianenigma commented on frame/elections-phragmen/src/benchmarking.rs:
```diff
@@ -520,50 +415,50 @@ benchmarks! {
+		// being set to max and this is a problem. In these cases, we cap e to a lower value, namely
+		// v * MAXIMUM_VOTE. when e is being iterated, v is at max, and again fine.
+		// all in all, votes_per_voter can never be more than MAXIMUM_VOTE.
+		let votes_per_voter = (e / v).min(MAXIMUM_VOTE as u32);
```
When v is being iterated e is always going to be MAX_VOTES, meaning that as we increase the number of voters, number of `edges_per_voter` stays the same (16 = MAX_VOTES) but the total number of edges will be variable. While it is not possible to guess if it is insignificant, I think this is the best that we can do with the current benchmarks. 

> Maybe a cleaner solution could be to have 2 benchmarks on which computes cost of c and e with fixed v, and another with fixed c and e, and varying v.

Yes this seems to be more accurate. The hassle will be that these weights will need to be applied manually though. I will see try and see if it gives different results.

kianenigma replied:
Here we have some results: 

```
	fn election_phragmen(c: u32, v: u32, e: u32, ) -> Weight {
		(0 as Weight)
			.saturating_add((42_535_000 as Weight).saturating_mul(c as Weight))
			.saturating_add((60_156_000 as Weight).saturating_mul(v as Weight))
			.saturating_add((4_149_000 as Weight).saturating_mul(e as Weight))
			.saturating_add(T::DbWeight::get().reads((2 as Weight).saturating_mul(c as Weight)))
			.saturating_add(T::DbWeight::get().reads((1 as Weight).saturating_mul(v as Weight)))
			.saturating_add(T::DbWeight::get().writes((1 as Weight).saturating_mul(c as Weight)))
	}
	fn election_phragmen_c_e(c: u32, e: u32, ) -> Weight {
		(0 as Weight)
			.saturating_add((47_360_000 as Weight).saturating_mul(c as Weight))
			.saturating_add((4_267_000 as Weight).saturating_mul(e as Weight))
			.saturating_add(T::DbWeight::get().reads(491 as Weight))
			.saturating_add(T::DbWeight::get().reads((2 as Weight).saturating_mul(c as Weight)))
			.saturating_add(T::DbWeight::get().writes((1 as Weight).saturating_mul(c as Weight)))
	}
	fn election_phragmen_v(v: u32, ) -> Weight {
		(3_522_562_000 as Weight)
			.saturating_add((7_101_000 as Weight).saturating_mul(v as Weight))
			.saturating_add(T::DbWeight::get().reads(386 as Weight))
			.saturating_add(T::DbWeight::get().reads((1 as Weight).saturating_mul(v as Weight)))
			.saturating_add(T::DbWeight::get().writes(186 as Weight))
	}

```

And it indeed does seem like v is being overestimated. Recall that this weight is basically is only used in on_initialize, and it was used to be set to `max_value()`. I would, for now mark the new benchmarks that break this down as `#[extra]`, leave the  `election_phragmen` to be the one with slight overestimation, and make an issue for it.

thiolliere replied:
having a comment stating what you just said and maybe an issue link it good to me 👍

thiolliere commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -129,79 +129,118 @@ type NegativeImbalanceOf<T> =
+	/// Be aware that this migration is intended to be used only for the mentioned versions. Use
+	/// with care and run at your own risk.
+	pub fn apply<
+		Instance: StorageInstance,
```
`StorageInstance` holds the information of one storage. I.e. module prefix and storage prefix.
But here you use it for multiple storage.

For now both module_prefix and storage_prefix are fixed inside elections-phragmen so we shouldn't need to give any of these information.
Or we can give the argument `module_prefix`.

thiolliere commented:
I made struct in migration private, added some doc and merged master

thiolliere reviewed (approved):
I can't really approve migrations_3_0_0.rs file as I participated in its writing.
But otherwise all looks good to me

Note about migration design: the migrations logic is completely isolated from pallet code, thus there is no burden associated to its maintainance (except when frame-support types are changing, but they must be rare and not too difficult to upgrade).
Also it allows for chain to trigger old migration even when types such as Balance have changed.

shawntabrizi commented on frame/elections-phragmen/src/migrations_3_0_0.rs:
```diff
@@ -0,0 +1,195 @@
+///
+/// ### Warning
+///
+/// This code will **ONLY** check that the storage version is less than 2_0_0. Further check
```
```suggestion
/// This code will **ONLY** check that the storage version is less than or equal to 2_0_0. Further check
```

shawntabrizi commented on frame/elections-phragmen/src/migrations_3_0_0.rs:
```diff
@@ -0,0 +1,195 @@
+///
+/// ### Warning
+///
+/// This code will **ONLY** check that the storage version is less than 2_0_0. Further check
+/// might be needed at the user runtime.
```
```suggestion
/// This code will **ONLY** check that the storage version is less than or equal to 2_0_0.
/// Further check might be needed at the user runtime.
```

kianenigma commented on frame/elections-phragmen/src/lib.rs:
```diff
@@ -289,210 +324,160 @@ decl_error! {
 
 			// Amount to be locked up.
-			let locked_balance = value.min(T::Currency::total_balance(&who));
+			let locked_stake = value.min(T::Currency::total_balance(&who));
```
I want someone to correct me here: Is this sane? Essentially, you can still lock up tokens that you have reserved in the system for voting, and you can seemingly use them for transaction fees. I would change that to `WithdrawReasons::all()` cc @shawntabrizi @thiolliere

thiolliere replied:
I can't really see why locking only for transaction-payment reason here. But I can't really say about those economic constraints/incentives.

kianenigma replied:
> I can't really see why locking only for transaction-payment reason here.

FYI we do the opposite, locking except for transaction payment, so you can pay for your fees with the vote stake that you've locked. Maybe @gavofyork can chime in? Else I would make this `::all()`, and change the doc to demonstrate.

kianenigma replied:
As promised, I updated this.

gavofyork replied:
Transaction fees mostly go into the system treasury (the same place as any slashes) so I'd generally consider them safe to spend out of a democracy vote (unlike e.g. tips). But it's mostly just to avoid the papercut of locking all your funds in a vote and then not having the funds to pay the fee to unlock them. It's probably fine as wallets/infra begins to understand stuff like this and automatically keeps a little back from being locked, to pay for any future transaction fees.

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

kianenigma commented:
This is queued for Runtime release 28 -- 27 is already drafted and does not contain this and will be launched next year. So Icing this in the meantime.

kianenigma commented:
/benchmark runtime pallet pallet_elections_phragmen

gavofyork commented:
Needs a companion?

kianenigma commented:
bot merge
